static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+	if (data.slotId) {
+		const container = document.getElementById(`dfp-ad--${data.slotId}`);
+		const iframes = container
+			? container.getElementsByTagName('iframe')
+			: null;
+		return iframes?.length ? iframes[0] : undefined;
+	} else if (data.iframeId) {
+		const iframe = document.querySelector<HTMLIFrameElement>(
+			`iframe#${data.iframeId}`,
+		);
+		return iframe ?? undefined;
+	}
+};
+
+// Regex for testing validity of message ids
+const validMessageRegex = /^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/;
+
+/**
+ * Narrow an `unknown` payload to the standard message format
+ *
+ * Until DFP provides a way for us to identify with 100% certainty our
+ * in-house creatives, we are left with doing some basic tests
+ * such as validating the anatomy of the payload and whitelisting
+ * event type
+ */
+const isValidPayload = (payload: unknown): payload is StandardMessage => {
+	const payloadToCheck = payload as StandardMessage;
+	return (
+		'type' in payloadToCheck &&
+		'value' in payloadToCheck &&
+		'id' in payloadToCheck &&
+		payloadToCheck.type in LISTENERS &&
+		validMessageRegex.test(payloadToCheck.id)
+	);
+};
+
+/**
+ * Cheap string formatting function
+ *
+ * @param error An object `{ code, message }`. `message` is a string where successive
+ * occurrences of %% will be replaced by the following arguments
+ * @param args Arguments that will replace %%
+ *
+ * @example
+ * formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+ * => { message: "Regis, you are so lovely" }
+ */
+const formatError = (
+	error: { code: number; message: string },
+	...args: string[]
+) =>
+	args.reduce((e, arg) => {
+		e.message = e.message.replace('%%', arg);
+		return e;
+	}, error);
+
+/**
+ * Convert a posted message to our StandardMessage format
+ *
+ * @param event The message event received on the window
+ * @returns A message with the `StandardMessage` format, or null if the conversion was unsuccessful
+ */
+const eventToStandardMessage = (event: MessageEvent) => {
+	try {
+		// Currently all non-string messages are discarded here since parsing throws an error
+		// TODO Review whether this is the desired outcome
+		const data: unknown = JSON.parse(event.data);
+		return isProgrammaticMessage(data)
+			? toStandardMessage(data)
+			: isValidPayload(data)
+			? data
+			: null;
+	} catch (ex) {
+		return null;
+	}
+};
+
+/**
+ * Respond to the original iframe with the result of calling the
+ * persistent listener / listener chain
+ */
+const respond = (
+	id: string,
+	target: MessageEventSource | null,
+	error: { message: string } | null,
+	result: unknown,
+) => {
+	postMessage(
+		{
+			id,
+			error,
+			result,
+		},
+		target ?? window,
+	);
+};
+
+/**
+ * Callback that is fired when an arbitrary message is received on the window
+ *
+ * @param event The message event received on the window
+ */
+const onMessage = async (event: MessageEvent<string>): Promise<void> => {
+	const data = eventToStandardMessage(event);
```

Comment:
Subjective but I prefer `data` -> `message` to align with the rest of the naming.